### ISSUE TITLE
'cube_build' crashes with the default 'band' argument ('all')

### ISSUE THREAD

**Issue by mcara** [STScI-JWST/jwst#830]
When building a cube using 'all' bands (which is the default value), cube builder crashes with the following error:
```python
>>> from jwst import cube_build
>>> c = cube_build.cube_build_step.CubeBuildStep.call('MIRM103-Q0-SHORT_495_rate_bsub_cal.fits', channel='1', band='all')
2017-04-18 14:03:45,297 - stpipe.CubeBuildStep - INFO - CubeBuildStep instance created.
2017-04-18 14:03:46,531 - stpipe.CubeBuildStep - INFO - Step CubeBuildStep running with args ('MIRM103-Q0-SHORT_495_rate_bsub_cal.fits',).
2017-04-18 14:03:46,531 - stpipe.CubeBuildStep - INFO - Starting IFU Cube Building Step
2017-04-18 14:03:46,531 - stpipe.CubeBuildStep - INFO - Input interpolation: pointcloud
2017-04-18 14:03:46,531 - stpipe.CubeBuildStep - INFO - Coordinate system to use: ra-dec
2017-04-18 14:03:46,531 - stpipe.CubeBuildStep - INFO - Weighting method for point cloud: standard
2017-04-18 14:03:46,531 - stpipe.CubeBuildStep - INFO - Power Weighting distance : 2.000000
2017-04-18 14:03:58,626 - stpipe.CubeBuildStep - INFO - Output Base MIRM103-Q0-SHORT_495_rate_bsub_cal 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../jwst/stpipe/step.py", line 458, in call
    return instance.run(*args)
  File "/.../jwst/stpipe/step.py", line 352, in run
    result = self.process(*args)
  File "/.../jwst/cube_build/cube_build_step.py", line 163, in process
    cube_build_io.DetermineCubeCoverage(self, MasterTable)
  File "/.../jwst/cube_build/cube_build_io.py", line 207, in DetermineCubeCoverage
    raise ErrorMissingParameter("Channel specified, but Subchannel was not")
jwst.cube_build.cube_build_io.ErrorMissingParameter: Channel specified, but Subchannel was not
```
I would assume that subchannel need not be set (by me) when ``band`` is set to ``'all'``: the code should figure out the subchannel itself.

CC: @jemorrison

**Comment by jemorrison** [comment]
Mihia
Currently there is a 1-1 mapping between channel and band
use --channel =1 --band = short or whatever

I will make the option 'all' more clear in the future

**Comment by mcara** [comment]
@jemorrison wrote:
> Currently there is a 1-1 mapping between channel and band

What do you mean by "1-1" mapping? What is the "all" option for?

CC: @hbushouse 
How can we build a cube from 2D images in _different_ bands?

**Comment by jemorrison** [comment]
Channel = 'All' and band = 'All' go together.
If the user gives nothing then the default is Channel='ALL' and band = 'ALL'
but once the user selects a channel to use like channel = 1 then band 
has to be set to either short,medium, or long.

This is mainly because of NIRSPEC. If the Grating is set then the Filter 
must be set.

Ok for MIRI this is a bit confusing - let me think about something that 
makes it more straight forward.

Jane

**Comment by mcara** [comment]
Indeed, the following works:
```python
>>> from jwst import datamodels, cube_build
>>> m = datamodels.ImageModel('MIRM103-Q0-SHORT_495_rate_bsub_cal.fits')
>>> c = cube_build.cube_build_step.CubeBuildStep.call(m, channel='ALL', band='ALL')
```
However, it seems strange (to me) that I need to worry about setting a specific band when I need either 'ALL' or probably 'ANY' for a given channel.

**Comment by jemorrison** [comment]
Strange to you but not to me. But I will make it work for you.

I am struggling with the case where the user may have an association 
table consisting
of all 4 channels and 3 bands. By defaults one huge UBER cube it made. 
But maybe they
want (and they will want to ) create a cube from 1 channel and 1 band. I 
wanted the code
to be very clear what the IFUcube is - so if a user selected a channel 
they also had to select the
band. But I can see that logic is tripping you up and will likely trip 
up others - so I am fixing it.
Jane

**Comment by mcara** [comment]
@jemorrison Many thanks!

**Comment by hbushouse** [comment]
@jemorrison which PR addresses this issue?

**Comment by jemorrison** [comment]
Mihai
When I tested this I just did not use --band.
So do not use --band = ALL - leave off the --band all together if you want
to use all the
bands.
I will go back and look at the logic again on how this works

Jane

**Comment by mcara** [comment]
@jemorrison wrote:
> When I tested this I just did not use --band. So do not use --band = ALL - leave off the --band all together if you want to use all the bands.

This is a strange advice that I do not believe can work at all. As the title of this issue says: _"'cube_build' crashes with the **default** 'band' argument ('all')"_ That is, the default value (=if not explicitly set by me) of the ``band`` parameter is ``'all'`` so I do not see how specifying it or not could affect the outcome.

From the [code](https://github.com/STScI-JWST/jwst/blob/master/jwst/cube_build/cube_build_step.py#L33) we can see that indeed ``'all'`` is the default value for the band parameter:
```python
class CubeBuildStep (Step):
    """
    CubeBuildStep: Creates a 3-D spectral cube from a given association or single input file
    The association will contain which channel/subchannel (MIRI) or filter/grating (NIRSPEC)
    the IFU Cube is going to cover. 
    """

    spec = """
         channel = option('1','2','3','4','ALL','all',default='ALL')
         band = option('SHORT','MEDIUM','LONG','ALL','short','medium','long','all',default='ALL')
```

**Comment by mcara** [comment]
It would be useful to link here the updates that made the code work, at least partially. In fact now the code works (at least for a longer time before crashing) when band is set to 'ALL' or not specified at all (as expected because the default value is 'ALL'). That is, the code when I reported this problem was crashing regardless of me explicitly (or not) setting band='ALL'. Now it does not crash with 'Channel specified, but Subchannel was not' error regardless of specifying ``band`` argument. That is, there must be a change to the code that was made to fix this issue.

However, it now crashes with a different error:
```python
>>> c = cube_build.cube_build_step.CubeBuildStep.call('MIRM103-Q0-SHORT_495_rate_bsub_cal.fits', channel='1', band='ALL')
2017-05-31 11:43:49,236 - stpipe.CubeBuildStep - INFO - CubeBuildStep instance created.
2017-05-31 11:43:49,770 - stpipe.CubeBuildStep - INFO - Step CubeBuildStep running with args ('MIRM103-Q0-SHORT_495_rate_bsub_cal.fits',).
2017-05-31 11:43:49,770 - stpipe.CubeBuildStep - INFO - Starting IFU Cube Building Step
2017-05-31 11:43:49,770 - stpipe.CubeBuildStep - INFO - Input interpolation: pointcloud
2017-05-31 11:43:49,771 - stpipe.CubeBuildStep - INFO - Coordinate system to use: ra-dec
2017-05-31 11:43:49,771 - stpipe.CubeBuildStep - INFO - Weighting method for point cloud: standard
2017-05-31 11:43:49,771 - stpipe.CubeBuildStep - INFO - Power Weighting distance : 2.000000
2017-05-31 11:43:59,665 - stpipe.CubeBuildStep - INFO - Output Base MIRM103-Q0-SHORT_495_rate_bsub_cal 
2017-05-31 11:44:00,766 - stpipe.CubeBuildStep - INFO - The desired cubes covers the MIRI Channels: ['1']
2017-05-31 11:44:00,766 - stpipe.CubeBuildStep - INFO - The desired cubes covers the MIRI subchannels: ['SHORT']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/.../jwst/stpipe/step.py", line 464, in call
    return instance.run(*args)
  File "/Users/.../jwst/stpipe/step.py", line 355, in run
    result = self.process(*args)
  File "/Users/.../jwst/cube_build/cube_build_step.py", line 201, in process
    self.par_filename = self.get_reference_file(self.models[0], 'cubepars')
  File "/Users/.../jwst/stpipe/step.py", line 618, in get_reference_file
    input_file, reference_file_type)
  File "/Users/.../jwst/stpipe/crds_client.py", line 132, in get_reference_file
    return list(get_multiple_reference_paths(input_file, [reference_file_type]).items())[0][1]
  File "/Users/.../jwst/stpipe/crds_client.py", line 88, in get_multiple_reference_paths
    bestrefs = crds.getreferences(data_dict, reftypes=reference_file_types, observatory="jwst")
  File "/Users/.../lib/python2.7/site-packages/crds/core/heavy_client.py", line 95, in getreferences
    parameters, reftypes, context, ignore_cache, observatory, fast)
  File "/Users/.../lib/python2.7/site-packages/crds/core/heavy_client.py", line 182, in _initial_recommendations
    parameters, reftypes=reftypes, context=final_context, ignore_cache=ignore_cache)
  File "/Users/.../lib/python2.7/site-packages/crds/core/heavy_client.py", line 316, in local_bestrefs
    return hv_best_references(context, parameters, reftypes)
  File "/Users/.../lib/python2.7/site-packages/crds/core/heavy_client.py", line 342, in hv_best_references
    minheader = ctx.minimize_header(conditioned)
  File "/Users/.../lib/python2.7/site-packages/crds/core/rmap.py", line 409, in minimize_header
    mapping = self.get_imap(instrument)
  File "/Users/.../lib/python2.7/site-packages/crds/core/rmap.py", line 745, in get_imap
    " for context " + repr(self.basename))
crds.core.exceptions.CrdsUnknownInstrumentError: Unknown instrument 'miri' for context 'hst_0526.pmap'
```
What am I doing wrong?

**Comment by mcara** [comment]
Also, it is worrying that the last log message before the crash says that the sub-channel is set to 'SHORT'... Is 'ALL'<=>'SHORT' ???
```python
>>> c = cube_build.cube_build_step.CubeBuildStep.call('MIRM103-Q0-SHORT_495_rate_bsub_cal.fits', channel='1')
2017-05-31 11:50:18,781 - stpipe.CubeBuildStep - INFO - CubeBuildStep instance created.
2017-05-31 11:50:19,591 - stpipe.CubeBuildStep - INFO - Step CubeBuildStep running with args ('MIRM103-Q0-SHORT_495_rate_bsub_cal.fits',).
2017-05-31 11:50:19,591 - stpipe.CubeBuildStep - INFO - Starting IFU Cube Building Step
2017-05-31 11:50:19,591 - stpipe.CubeBuildStep - INFO - Input interpolation: pointcloud
2017-05-31 11:50:19,591 - stpipe.CubeBuildStep - INFO - Coordinate system to use: ra-dec
2017-05-31 11:50:19,591 - stpipe.CubeBuildStep - INFO - Weighting method for point cloud: standard
2017-05-31 11:50:19,591 - stpipe.CubeBuildStep - INFO - Power Weighting distance : 2.000000
2017-05-31 11:50:30,046 - stpipe.CubeBuildStep - INFO - Output Base MIRM103-Q0-SHORT_495_rate_bsub_cal 
2017-05-31 11:50:31,028 - stpipe.CubeBuildStep - INFO - The desired cubes covers the MIRI Channels: ['1']
2017-05-31 11:50:31,028 - stpipe.CubeBuildStep - INFO - The desired cubes covers the MIRI subchannels: ['SHORT']
```

**Comment by hbushouse** [comment]
You're using the wrong CRDS pmap: "crds.core.exceptions.CrdsUnknownInstrumentError: Unknown instrument 'miri' for context 'hst_0526.pmap'". This one is obviously for HST. You need the JWST pmap. Try setting:

export CRDS_CONTEXT='jwst_0332.pmap'

**Comment by hbushouse** [comment]
You've only provided a single input exposure "MIRM103-Q0-SHORT_495_rate_bsub_cal.fits", which can only contain data for a single band. In this case it appears to be SHORT. Data for other bands will be contained in other exposures, so if you want cubes built for multiple bands, it will be necessary to supply multiple inputs.

Remember: 1 MIRI MRS exposure/file contains data for 2 channels and 1 band.

**Comment by mcara** [comment]
> Remember: 1 MIRI MRS exposure/file contains data for 2 channels and 1 band.
Thank you!

**Comment by jemorrison** [comment]
Closed

**Comment by jemorrison** [comment]
Closed #830 via #939
